src/testRunner/unittests/tsbuild/noCheck.ts

```diff
@@ -0,0 +1,86 @@
+import {
+    CommandLineOption,
+    optionDeclarations,
+} from "../../_namespaces/ts";
+import { jsonToReadableText } from "../helpers";
+import {
+    noChangeRun,
+    verifyTsc,
+} from "../helpers/tsc";
+import { loadProjectFromFiles } from "../helpers/vfs";
+
+function verifyNoCheckFlag(variant: string) {
+    function verifyNoCheckWorker(subScenario: string, declAText: string, commandLineArgs: readonly string[]) {
+        verifyTsc({
+            scenario: variant,
+            subScenario,
+            fs: () =>
+                loadProjectFromFiles({
+                    "/src/a.ts": getATsContent(declAText),
+                    "/src/tsconfig.json": jsonToReadableText({
+                        compilerOptions: { noCheck: true, emitDeclarationOnly: true, declaration: true },

```

Comment:
Having this in json and not reporting error is a bug because then it doesnt become API only option per design meeting.,

The scenario i am talking is missed here because:
modifying tsconfig.json will check that config file is newer than output and hence program will be rebuilt. The problem with doing this with API is that output will be updated but not inputs which means the build will report that everything is uptodate

Here is how to write this test case
1 load project files (without noCheck option) in your test case as is . No errors
2. As part of first edit: Do these steps to emulate API usage:  Change a.ts to introduce error, run API (by creating program and emitting with noCheck on, use `sys= new vfs.sys(fs))` to create compiler host  
3. This will fail because after edit, test framework will run `tsc --build` again and it will find that everything is UpToDate and not report error.  (test will not create discrepancy baseline because unfortunately we don't check discrepancies in error reporting just file emit validation)

Reply:
> Having this in json and not reporting error is a bug because then it doesnt become API only option per design meeting.,

And it *does* _normally_ report an error, which is what the other test block that doesn't override the validator function verifies (and naturally, these collapse into one test once it's actually exposed via CLI)~

> The problem with doing this with API is that output will be updated but not inputs which means the build will report that everything is uptodate

If you build via API and pass options not in your on-disk config file (as you would need to, since `noCheck` is API only right now), your up-to-dateness checks for configs are going to *all* be invalidated anyway, since your API-fueled build will have different options applied than a commandline one. Precisely

> The problem with doing this with API is that output will be updated but not inputs which means the build will report that everything is uptodate

is in no way unique to this new flag - any time you do a API build with different flags than what your on-disk configs specify this'll happen, and I'm pretty sure how `-b` handles that is just entirely out of scope, unless we're talking specifically `incremental` output, which just comes down to the options serialized in the `.buildinfo` which is why:

> This will fail because after edit, test framework will run tsc --build again and it will find that everything is UpToDate and not report error. (test will not create discrepancy baseline because unfortunately we don't check discrepancies in error reporting just file emit validation)

It both will and won't. If `--incremental` isn't on, yeah, it won't build again because all the timestamps say it's newer, and `build` doesn't know that your last build was with different settings. Again, that's pretty out-of-scope and isn't going to get fixed, since there's no incremental artifact. Nothing new there, that issue has nothing to do with `noCheck` - you can replicate that with any semantic diagnostic affecting option. For `--incremental`, that *won't* be the case, because the `.buildinfo` emitted by the API build will have different compiler options than the new build (semantic diagnostic affecting ones), which'll force a recheck (though not reemit), and is _precisely_ what this test verifies without needing all the API-layering hoopla by temporarily enabling it in configs/cli (since an incremental build via config is going to produce the same buildinfo as one via api). AFAIK, the only critical thing to test is that `noCheck` is serialized into the `buildinfo` just like every other semantic diagnostic-affecting option, when it's allowed, which this test verifies.

As an aside "via API" is ambiguous, since we expose `executeCommandLine` now, but I've taken that to mean "only allowed where we take `CompilerOptions`-type objects" for now, which just means any API that's post-option-parsing/validating. Absolutely none of those APIs (eg, `createProgram`) are even going to emit a `.buildinfo` file on their own without extra instrumentation on behalf of the caller of the API, so it is quite literally up to the API user to keep things up-to-date, provided our `emitBuildInfo` function does serialize the option for them when they call it to do that, which, again, this test does verify.

Reply:
The problem with writing this test case this way is that it is modifying timestamp of tsconfig files and hence build will take place. To check this correctly, you want to use API emit as `edit` as mentioned above.

Reply:
Or I could allow `--noCheck` as a `build` option (why not? seems like a reasonable thing to want to turn on/off for the whole project graph) and pass it down from the top-level CLI invocation, rather than by editing the `tsconfig`?